syntax = "proto3";
package core.protos.playbooks;

import "google/protobuf/wrappers.proto";

message Bitbucket {
  message FetchFile {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
    google.protobuf.StringValue branch = 4;
    sfixed64 timestamp = 5;
  }

  message UpdateFile {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
    google.protobuf.StringValue content = 4;
    google.protobuf.StringValue commit_message = 5;
    google.protobuf.StringValue branch = 6;
    google.protobuf.StringValue author = 7;
  }

  message FetchCommits {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue branch = 3;
    google.protobuf.StringValue author = 4;
  }

  message FetchRelatedCommits {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
    google.protobuf.StringValue function_name = 4;
    google.protobuf.StringValue branch = 5;
  }

  message FetchRecentMerges {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue branch = 3;
  }

  message FileUpdate {
    google.protobuf.StringValue path = 1;
    google.protobuf.StringValue content = 2;
  }

  message CreatePullRequest {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue title = 3;
    google.protobuf.StringValue description = 4;
    google.protobuf.StringValue source_branch = 5;
    google.protobuf.StringValue destination_branch = 6;
    repeated FileUpdate files = 7;
    google.protobuf.StringValue commit_message = 8;
    google.protobuf.BoolValue close_source_branch = 9;
  }

  message FetchPullRequestDiff {
    google.protobuf.StringValue workspace = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.Int64Value pr_id = 3;
  }

  enum TaskType {
    UNKNOWN = 0;
    FETCH_FILE = 1;
    UPDATE_FILE = 2;
    FETCH_COMMITS = 3;
    FETCH_RELATED_COMMITS = 4;
    FETCH_RECENT_MERGES = 5;
    CREATE_PULL_REQUEST = 6;
    FETCH_PULL_REQUEST_DIFF = 7;
  }

  TaskType type = 1;
  oneof task {
    FetchFile fetch_file = 2;
    UpdateFile update_file = 3;
    FetchCommits fetch_commits = 4;
    FetchRelatedCommits fetch_related_commits = 5;
    FetchRecentMerges fetch_recent_merges = 6;
    CreatePullRequest create_pull_request = 7;
    FetchPullRequestDiff fetch_pull_request_diff = 8;
  }
}
